docs/root/overview.md

```diff
@@ -0,0 +1,189 @@
+per-phase reporting of latencies as well as counters and latencies.
+
+Concretely, a warm-up phase could be represented by a distinct duration
+termination predicate and a ramping rate limiter. Upon completion, the `hot`
+BenchmarkClient with its associated pool would then be live-transferred to the
+next configured phase, after which execution can continue.
+
+One other reason to have this is that it would enable remote- and/or cli-
+controlled ramping of certain test parameters by associating those to different
+phases. Termination predicates can be leveraged to immediately terminate the
+current phase after injecting a new one, allowing for real-time steering via
+gRPC and/or CLI.
+
+### Streaming parameterization and output stats
+
+Once we have phases, the gRPC service, and perhaps the CLI, would be natural
+candidates to follow up with to allow dynamic phase injection, as well as send
+back reports per phase.
+
+## Key concept descriptions
+
+*The c++ interface definitions for the concepts below can be found [here](https://github.com/envoyproxy/nighthawk/tree/master/include/nighthawk)*.
+
+### Process
+
+**Process** represents the primary entry point to a Nighthawk execution run.
+Only one Process is allowed to exist at the same point in time within an OS
+Process. **Process** is responsible for performing process-wide initialization
+and termination, as well as handle input configuration, deliver output, and
+co-ordinate Workers. **ProcessImpl** is re-used across the CLI and the gRPC
+service.
+
+### Worker
+
+**Worker** is responsible for performing correct initialization and termination
+of its thread, as well as execution of its designated task and offering a way
+for consumers to wait for that task to complete.
+
+### TerminationPredicate
+
+**TerminationPredicate** is responsible for if and how to terminate. As of
+today, there are two types: one that will indicate that it is time to terminate
+based on a pre-configured duration, and one that will do so based on absolute
+counter thresholds.
+
+### Sequencer
+
+**SequencerImpl** resides on a worker-thread, and drives itself via timers that
+run on the dispatcher, and coordinates interaction between **RateLimiter,**
+**BenchmarkClient**, and **TerminationPredicate** to drive execution to
+completion.
+
+### RateLimiter
+
+**RateLimiter** is responsible for indicating when it is time to release a
+request. **RateLimiter** offers a semaphore-like interaction model, as in
+[closed-loop](terminology.md#closed-loop) mode it may be that
+**BenchmarkClient** is not able to satisfy request-release timings, in which
+case acquisitions from **RateLimiter** need to be cancelled. Concretely, as of
+today there is **LinearRateLimiterImpl** which offers a straight-paced plain
+frequency, as well as work in progress on
+**DistributionSamplingRateLimiterImpl** (adding uniformly distributed random
+timing offsets to an underlying **RateLimiter**) and **RampingRateLimiter**.
```

Comment:
I think we ended up calling it LinearRampingRateLimiter instead of RampingRateLimiter.